api/v1beta2/ibmpowervscluster_webhook.go

```diff
@@ -97,7 +98,42 @@ func (r *IBMPowerVSCluster) validateIBMPowerVSClusterNetwork() *field.Error {
 	return nil
 }
 
-func (r *IBMPowerVSCluster) validateIBMPowerVSClusterCreateInfraPrereq() *field.Error {
+func (r *IBMPowerVSCluster) validateIBMPowerVSClusterLoadBalancerPort() (allErrs field.ErrorList) {
+	for i, loadbalancer := range r.Spec.LoadBalancers {
+		for j, listerner := range loadbalancer.AdditionalListeners {
+			if listerner.Port == int64(DefaultAPIServerPort) {
```

Comment:
what if this port is set `Cluster.Spec.ClusterNetwork.APIServerPort` ? will this comparison with `DefaultAPIServerPort` still hold good?

Reply:
No, it won't hold good, if user has set `Cluster.Spec.ClusterNetwork.APIServerPort` we should make sure that that port is not set here in listner.
We usually find the apiserver port like this https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/81e93e2aa11f6ce94ff4df941720b02224f812d4/cloud/scope/powervs_cluster.go#L336-L342

Reply:
we dont have access to cluster resource here in webhook, So we need verify this one scenario in controller code only.

Reply:
then adding this check here for the `DefaultAPIServerPort` may mislead the user, lets drop this change here and introduce in the controller itself

Reply:
I think most of the time the apiserver will be running in 6443 only, so i think we can keep it here and can add a check in controller only if the apiserver port is not 6443.

Reply:
Yeah, Agree but only thing is in controller the check is performed on every reconcile where as here its once. Thats the  only concern I have.

Reply:
you are right, lets not add here for the sake of adding if its not 100% accurate.

Reply:
sure @Karthik-K-N. I will make the necessary changes.

Reply:
When this check is moved to controller, this check is failing on every reconcile and clusters is in provisioning state.
```
[manager] E0522 09:16:02.569067       1 controller.go:329] "Reconciler error" err="loadbalancer capi-powervs-keerthana-new-loadbalancer2.AdditionalListerners[0].Port value should not be equal to (6443) as Port (6443) is already configured as apiserverport to loadbalancer" controller="ibmpowervscluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="IBMPowerVSCluster" IBMPowerVSCluster="default/capi-powervs-keerthana-new" namespace="default" name="capi-powervs-keerthana-new" reconcileID="0f6cfb1c-49e3-4272-b17a-7b880fb400b4"
[manager] E0522 09:16:17.635519       1 controller.go:329] "Reconciler error" err="loadbalancer capi-powervs-keerthana-new-loadbalancer2.AdditionalListerners[0].Port value should not be equal to (6443) as Port (6443) is already configured as apiserverport to loadbalancer" controller="ibmpowervscluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="IBMPowerVSCluster" IBMPowerVSCluster="default/capi-powervs-keerthana-new" namespace="default" name="capi-powervs-keerthana-new" reconcileID="1382c88c-4355-421e-9bcc-0cb38917117d"
[manager] E0522 09:16:31.443073       1 controller.go:329] "Reconciler error" err="loadbalancer capi-powervs-keerthana-new-loadbalancer2.AdditionalListerners[0].Port value should not be equal to (6443) as Port (6443) is already configured as apiserverport to loadbalancer" controller="ibmpowervscluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="IBMPowerVSCluster" IBMPowerVSCluster="default/capi-powervs-keerthana-new" namespace="default" name="capi-powervs-keerthana-new" reconcileID="73fed8f4-f017-49ac-bb72-cf879789899c"
[manager] E0522 09:16:48.998575       1 controller.go:329] "Reconciler error" err="loadbalancer capi-powervs-keerthana-new-loadbalancer2.AdditionalListerners[0].Port value should not be equal to (6443) as Port (6443) is already configured as apiserverport to loadbalancer" controller="ibmpowervscluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="IBMPowerVSCluster" IBMPowerVSCluster="default/capi-powervs-keerthana-new" namespace="default" name="capi-powervs-keerthana-new" reconcileID="d45f74d4-09b0-48f5-91db-18a7555ad20a"
```

Reply:
I think only logs might not help, either we need to set some conditions or events. Eventually user needs to edit the spec or delete and reapply with modified spec.

Reply:
if this is getting called in `ReconcileLoadBalancer` method, don't we update the condition for the load balancer with these errors?

Reply:
Yes, we do set here https://github.com/damdo/cluster-api-provider-ibmcloud/blob/fd15e31a193d0b58a186c70121b7d3b559235ce9/controllers/ibmpowervscluster_controller.go#L215-L219, I think this should be good for user to understand. May be @KeerthanaAP could you please paste the output of IBMPowerVSCluster object's condition in case of failure.

Reply:
```
Conditions:
    Last Transition Time:  2024-05-27T09:42:51Z
    Message:               loadbalancer capi-powervs-keerthana-new-loadbalancer2.AdditionalListerners[0].Port value should not be equal to (6443) as Port (6443) is already configured as apiserverport to loadbalancer
    Reason:                LoadBalancerReconciliationFailed

```
Currently we are getting the above error in the conditions. will update the error as `capi-powervs-keerthana-new-loadbalancer2 port (6443) cannot be used as additional Listener port value as its already in use`

Reply:
Sure. Will add this as error msg.

Reply:
Modified the error msg.
Updated output: 
```
Conditions:
    Last Transition Time:  2024-05-27T17:01:33Z
    Message:               capi-powervs-keerthana-new-loadbalancer2 port (6443) cannot be used as additional Listener port value, as its already configured by system for default listener
    Reason:                LoadBalancerReconciliationFailed
```

Reply:
Nice, Lets hear others opinion on as well.

Reply:
how about this one - https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1746#discussion_r1618191581?

Reply:
Looks good. As the error strings should not be capitalized, Can we update it as below.
```
  Conditions:
    Last Transition Time:  2024-05-29T07:03:15Z
    Message:               port 6443 for the capi-powervs-keerthana-new-loadbalancer2 load balancer cannot be used as an additional listener port, as it is already assigned to the API server
    Reason:                LoadBalancerReconciliationFailed
```